### ISSUE TITLE
PubSub regression in 0.27

### ISSUE THREAD

**Issue by mattklein123** [redis-rs/redis-rs#1327]
I believe there is a regression in https://github.com/redis-rs/redis-rs/pull/1144 if using existing code and calling https://docs.rs/redis/latest/redis/aio/struct.PubSub.html#method.into_on_message. I haven't fully debugged yet but I think what is happening is that when the sink is dropped as part of this method, the backing task shuts down due to the sink sender going away and then no more messages are received.

I will see if I can work around it but if this is hard to fix and the new implementation is not going to be reverted this method should be removed.

**Comment by nihohit** [comment]
Can you give a test case reproducing this issue? We already have a test case that checks whether the `stream` can receive message after dropping the `sink` - https://github.com/redis-rs/redis-rs/blob/0f421bae30adb2428d011acaf06986648b63fd53/redis/tests/test_async.rs#L766
And I've tested locally, with this test that uses `into_on_message`, and the test passes, so I don't see the regression you describe.

```rs
#[test]
fn can_receive_messages_from_split_pub_sub_after_into_on_message() {
    let ctx = TestContext::new();
    block_on_all(async move {
        let mut pubsub = ctx.async_pubsub().await?;
        let mut publish_conn = ctx.async_connection().await?;

        let _: () = pubsub.subscribe("phonewave").await?;
        let mut stream = pubsub.into_on_message();
        let _: () = publish_conn.publish("phonewave", "banana").await?;

        let message: String = stream.next().await.unwrap().get_payload().unwrap();

        assert_eq!("banana".to_string(), message);

        Ok(())
    })
    .unwrap();
}
```

**Comment by mattklein123** [comment]
I can try to come up with something but I just changed my code to hold PubSub alive and use `on_message()` instead and it is back to working.

**Comment by mattklein123** [comment]
Don't know if helpful or not but here is a snippet of what I was doing before that stopped working:
```
async fn subscribe<'a, 'b>(&self, topic: Key<'b>) -> Result<BoxStream<'a, Result<String>>> {
    let client = Client::open(self.redis_host.as_str())?;
    let mut pubsub = client.get_async_pubsub().await?;
    pubsub.subscribe(topic).await?;

    // Now translate the incoming stream of messages into a stream of strings that can more easily
    // be mocked.
    Ok(Box::pin(
      pubsub
        .into_on_message()
        .map(|message| Ok(message.get_payload()?)),
    ))
  }
```

**Comment by nihohit** [comment]
It would be best if you can recreate the issue - you'd have your code as you want it, and other users won't have this issue.

**Comment by mattklein123** [comment]
I will see if I can come up with something self contained, not sure when I will get to it.

**Comment by nihohit** [comment]
I still can't repro, but if you can change 
`.map(|message| Ok(message.get_payload()?)),`
to 
```rs
.map(|message| {
  println!("{message:?}");
  let res = message.get_payload::<String>();
  println!("{res:?}");
  Ok(res?)
})
```

then maybe the prints will give us more information. Maybe the issue is that you're receiving some error that causes the stream to close - not that I understand why previously this didn't happen.

**Comment by nihohit** [comment]
I found the issue, it's fixed in version 0.27.2.
